### ISSUE TITLE
sort stupidness caused segfault

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4338]
Migrated from [rt.perl.org#7567](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7567) (status was 'resolved')

Searchable as RT7567$

**Comment by p5pRT** [comment]
### From root@ws.com.au



just stuffing around being stupid with sort \{\} and \.\.

gateway&#8203;:\~/tmp\# cat rand
\#\!/usr/bin/perl
@&#8203;a = \(1\, 2\, 3\, 4\, 5\, 6\, 7\, 8\, 9\);
@&#8203;a = sort \{
&nbsp;        @&#8203;a = sort \{
&nbsp;                @&#8203;a = sort \{
&nbsp;                        @&#8203;a = sort \{
&nbsp;                                int rand \(2\) \- 2;
&nbsp;                        \}
&nbsp;                        int rand \(2\) \- 2;
&nbsp;                \}
&nbsp;                int rand \(2\) \- 2;
&nbsp;        \} @&#8203;a;
&nbsp;        int rand \(2\) \- 2;
\} @&#8203;a;
print join ''\, @&#8203;a;
gateway&#8203;:\~/tmp\# \./rand
Segmentation fault

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.1:

Configured by root at Tue Apr 10 08:56:16 WST 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.2.17, archname=i686-linux
    uname='linux gateway 2.2.17 #3 tue oct 10 11:42:39 wst 2000 i686 unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.2 20000220 (Debian GNU/Linux)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -ldb -ldl -lm -lc -lposix -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.1.3.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /usr/lib/perl5/5.6.1/i686-linux
    /usr/lib/perl5/5.6.1
    /usr/lib/perl5/site_perl/5.6.1/i686-linux
    /usr/lib/perl5/site_perl/5.6.1
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/Debian
    .


Environment for perl v5.6.1:
    HOME=/root
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/scripts
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

I don't think this is directly related to sort because
a\) it happens both with the 5\.6 qsort and the 5\.7 merge sort and
b\) it goes away if you pass @&#8203;a to the two innermost invocations
&nbsp;   \(as is done on the outermost invocations\) before returning the
&nbsp;   random value\.

Still\, as the documentation says\,
&nbsp;   The comparison function is required to behave\.
A sort implementor should be able to pick any array element
and be confident that the comparison function will return 0
when the element is compared with itself\.  There are sort
implementations where the use of such sentinels lead to
significant savings\.  It happens that neither the qsort nor
the merge sort implementation rely on that\, so we're spared
the difficult choice of how much to penalize well\-behaved
users to spare bozos a core dump\.

We return you now to the hands of people better qualified to
determine the real cause of the core dump\.
\-\- jpl

**Comment by p5pRT** [comment]
### From @abigail

If you replace all the occurances of 'int rand \(2\) \- 2' in the above
code with 0\, it still core dumps\.

My guess is that the problem is the modification of @&#8203;a during the sort\.



Abigail

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

I agree\.  If it helps those better able to understand what goes on
before the sort is called\, and after the compare subroutine is called
\(the core certainly isn't bloated by excessive comments :\-\)

Program received signal SIGSEGV\, Segmentation fault\.
0x080d9c9a in Perl\_sv\_clear \(sv=0x81819a8\) at sv\.c&#8203;:4894
4894            if \(SvMAGIC\(sv\)\)
\(gdb\) where
\#0  0x080d9c9a in Perl\_sv\_clear \(sv=0x81819a8\) at sv\.c&#8203;:4894
\#1  0x080da299 in Perl\_sv\_free \(sv=0x81819a8\) at sv\.c&#8203;:5098
\#2  0x080f8e15 in Perl\_free\_tmps \(\) at scope\.c&#8203;:182
\#3  0x080c4b4a in Perl\_pp\_nextstate \(\) at pp\_hot\.c&#8203;:40
\#4  0x080c463b in Perl\_runops\_debug \(\) at run\.c&#8203;:53
\#5  0x0810c399 in sortcv \(a=0x8173bd4\, b=0x8173cb8\) at pp\_ctl\.c&#8203;:4183
\#6  0x0810bc86 in dynprep \(list1=0x81752cc\, list2=0x81818c8\, nmemb=9\, 
&nbsp;    cmp=0x810c348 \<sortcv>\) at pp\_ctl\.c&#8203;:3953
\#7  0x0810bfb4 in S\_qsortsv \(list1=0x81752cc\, nmemb=9\, cmp=0x810c348 \<sortcv>\)
&nbsp;    at pp\_ctl\.c&#8203;:4051
\#8  0x080feced in Perl\_pp\_sort \(\) at pp\_ctl\.c&#8203;:1026
\#9  0x080c463b in Perl\_runops\_debug \(\) at run\.c&#8203;:53
\#10 0x0805f92b in S\_run\_body \(oldscope=1\) at perl\.c&#8203;:1547
\#11 0x0805f51b in perl\_run \(my\_perl=0x8173a68\) at perl\.c&#8203;:1469
\#12 0x0805c47b in main \(argc=2\, argv=0xbffff8bc\, env=0xbffff8c8\)
&nbsp;    at perlmain\.c&#8203;:67
\#13 0x40083177 in \_\_libc\_start\_main \(main=0x805c400 \<main>\, argc=2\, 
&nbsp;    ubp\_av=0xbffff8bc\, init=0x805b4f0 \<\_init>\, fini=0x8153ccc \<\_fini>\, 
&nbsp;    rtld\_fini=0x4000e184 \<\_dl\_fini>\, stack\_end=0xbffff8b4\)
&nbsp;    at \.\./sysdeps/generic/libc\-start\.c&#8203;:129

I suspect\, after the innermost invocations of sort \(lacking the @&#8203;a
argument to sort\) cause @&#8203;a to shrink\, the references to the original
array elements in the outer invocation are no longer valid\.

At a philosophical level \(I don't know enough perlguts to determine
if it's practical\)\, it would be nice if the array that sort is
handed could\, in recent parlance\, be clamped\, so modifications
\(except by sort\, of course\, which is operating outside or deep
inside the language\, whichever you prefer\) would cause a fatal
error \(but not a coredump\)\.  I'm unable to think of any \*good\*
that can come from modifications to the array being sorted as
a side\-effect of the comparison routine\, and I can think of
dozens of \*bad\* things\.\.\. core dumps\, limitation on the sort
implementation\, increased overhead\, unexpected behavior\.

If the array can't be marked as read\-only\, I believe sortcv
could be fiddled \(by someone with more experience than I\)
to verify that a and b are valid references\.  But that will
slow down all sorts\, so it would be my second choice\.  \-\- jpl

**Comment by p5pRT** [comment]
### From @vanstyn

A while back\, "Abigail" \<abigail@&#8203;foad\.org> wrote&#8203;:
:On Wed\, Aug 22\, 2001 at 07&#8203;:17&#8203;:07AM \-0400\, John P\. Linderman wrote&#8203;:
:> > just stuffing around being stupid with sort \{\} and \.\.
:> > 
:> > gateway&#8203;:\~/tmp\# cat rand
:> > \#\!/usr/bin/perl
:> > @&#8203;a = \(1\, 2\, 3\, 4\, 5\, 6\, 7\, 8\, 9\);
:> > @&#8203;a = sort \{
:> >         @&#8203;a = sort \{
:> >                 @&#8203;a = sort \{
:> >                         @&#8203;a = sort \{
:> >                                 int rand \(2\) \- 2;
:> >                         \}
:> >                         int rand \(2\) \- 2;
:> >                 \}
:> >                 int rand \(2\) \- 2;
:> >         \} @&#8203;a;
:> >         int rand \(2\) \- 2;
:> > \} @&#8203;a;
:> > print join ''\, @&#8203;a;
:> > gateway&#8203;:\~/tmp\# \./rand
:> > Segmentation fault
:> 
:> I don't think this is directly related to sort because
:> a\) it happens both with the 5\.6 qsort and the 5\.7 merge sort and
:> b\) it goes away if you pass @&#8203;a to the two innermost invocations
:>    \(as is done on the outermost invocations\) before returning the
:>    random value\.
:> 
:> Still\, as the documentation says\,
:>    The comparison function is required to behave\.
:
:If you replace all the occurances of 'int rand \(2\) \- 2' in the above
:code with 0\, it still core dumps\.
:
:My guess is that the problem is the modification of @&#8203;a during the sort\.

This still dumps with latest bleadperl\. Here's a shorter test case&#8203;:
&nbsp;  \./perl \-wle '@&#8203;a=\(1\.\.8\); @&#8203;a = sort \{ @&#8203;a = \(0\) \} @&#8203;a'
\.\.\. and each occurrence of @&#8203;a appears to be required for the coredump\.

Hugo

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Hugo noted&#8203;:
> A while back\, "Abigail" \<abigail@&#8203;foad\.org> wrote&#8203;:
> :On Wed\, Aug 22\, 2001 at 07&#8203;:17&#8203;:07AM \-0400\, John P\. Linderman wrote&#8203;:
>
> This still dumps with latest bleadperl\. Here's a shorter test case&#8203;:
>   \./perl \-wle '@&#8203;a=\(1\.\.8\); @&#8203;a = sort \{ @&#8203;a = \(0\) \} @&#8203;a'
> \.\.\. and each occurrence of @&#8203;a appears to be required for the coredump\.
> 
> Hugo

I'm woefully ignorant of the setup prior to the sorting itself\.
\(So I can't offer any insight into the
 sub s1 \{ goto &s2 \} sub s2 \{ $a cmp $b \} print sort s1 \(3\, 2\, 1\)
problem\, either\)\.

I know that mergesort makes yet another copy of all the references
to the original list passed to sort\, as does the stabilized qsort\.
If the array is changed as part of the comparison routine\, these
references are going who\-knows\-where\.  Even if it is possible to
check on the validity of the reference\, it's not obvious \(to me\)
that the overhead \*all\* users would have to pay is worth the
protection provided to way\-out users\.

Should a warning against modifying the array as a side\-effect
of the comparison routine be added to the documentation?  \-\- jpl

**Comment by p5pRT** [comment]
### From @smpeters

> \[root@&#8203;ws\.com\.au \- Tue Aug 21 14&#8203;:53&#8203;:01 2001\]&#8203;:
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> just stuffing around being stupid with sort \{\} and \.\.
> 
> gateway&#8203;:\~/tmp\# cat rand
> \#\!/usr/bin/perl
> @&#8203;a = \(1\, 2\, 3\, 4\, 5\, 6\, 7\, 8\, 9\);
> @&#8203;a = sort \{
>         @&#8203;a = sort \{
>                 @&#8203;a = sort \{
>                         @&#8203;a = sort \{
>                                 int rand \(2\) \- 2;
>                         \}
>                         int rand \(2\) \- 2;
>                 \}
>                 int rand \(2\) \- 2;
>         \} @&#8203;a;
>         int rand \(2\) \- 2;
> \} @&#8203;a;
> print join ''\, @&#8203;a;
> gateway&#8203;:\~/tmp\# \./rand
> Segmentation fault
> 

This bug has been fixed by making it illegal to attempt to modify an array while sorting it\.  
This was fixed with change \#25953\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From jpl@research.att.com

> > just stuffing around being stupid with sort \{\} and \.\.
> > 
> > gateway&#8203;:\~/tmp\# cat rand
> > \#\!/usr/bin/perl
> > @&#8203;a = \(1\, 2\, 3\, 4\, 5\, 6\, 7\, 8\, 9\);
> > @&#8203;a = sort \{
> >         @&#8203;a = sort \{
> >                 @&#8203;a = sort \{
> >                         @&#8203;a = sort \{
> >                                 int rand \(2\) \- 2;
> >                         \}
> >                         int rand \(2\) \- 2;
> >                 \}
> >                 int rand \(2\) \- 2;
> >         \} @&#8203;a;
> >         int rand \(2\) \- 2;
> > \} @&#8203;a;
> > print join ''\, @&#8203;a;
> > gateway&#8203;:\~/tmp\# \./rand
> > Segmentation fault
> > 
> 
> This bug has been fixed by making it illegal to attempt
> to modify an array while sorting it\.  
> This was fixed with change \#25953\.

While it is rude to segfault\, it is also rude to ignore
the warnings in the manual&#8203;:

&nbsp;    The comparison function is required to behave\.  If it returns
&nbsp;    inconsistent results \(sometimes saying $x\[1\] is less than $x\[2\]
&nbsp;    and sometimes saying the opposite\, for example\) the results are
&nbsp;    not well\-defined\.

int rand \(2\) \- 2;

is not a valid comparison function\.  Current implementations of
the sort\(\) operation don't include segfaulting in the
"not well\-defined" consequences\.  But some perfectly fine
implementations could\.  For example\, I have seen partition
sorts that plunk the partitioning element in position 0\,
and rely on a 0 return to recognize that it might have
just been compared to itself\.  This can save an extra
comparison to check to see if the left end of the partition
has been reached\.  If the comparison function doesn't
return 0 when the pivot element is compared to itself\,
the sort will run off the end of the partition\, and almost
anything can happen\.  I don't like segfaults\, but I also
don't like penalizing well\-behaved users to prevent segfaults
for users that don't follow the rules\.  \-\- jpl

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Dec 14 18&#8203;:37&#8203;:46 2005\, stmpeters wrote&#8203;:
> > \[root@&#8203;ws\.com\.au \- Tue Aug 21 14&#8203;:53&#8203;:01 2001\]&#8203;:
> >
> >
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> >
> > just stuffing around being stupid with sort \{\} and \.\.
> >
> > gateway&#8203;:\~/tmp\# cat rand
> > \#\!/usr/bin/perl
> > @&#8203;a = \(1\, 2\, 3\, 4\, 5\, 6\, 7\, 8\, 9\);
> > @&#8203;a = sort \{
> >         @&#8203;a = sort \{
> >                 @&#8203;a = sort \{
> >                         @&#8203;a = sort \{
> >                                 int rand \(2\) \- 2;
> >                         \}
> >                         int rand \(2\) \- 2;
> >                 \}
> >                 int rand \(2\) \- 2;
> >         \} @&#8203;a;
> >         int rand \(2\) \- 2;
> > \} @&#8203;a;
> > print join ''\, @&#8203;a;
> > gateway&#8203;:\~/tmp\# \./rand
> > Segmentation fault
> >
> 
> This bug has been fixed by making it illegal to attempt to modify an
> array while sorting it\.
> This was fixed with change \#25953\.

\(aka 9850bf2\)

But that also caused the array to stay read\-only after an attempt to
modify it caused a croak\.  I fixed that in commit 20d5dc239d\.

\-\- 

Father Chrysostomos
